### ISSUE TITLE
DatabaseDateDecodingStrategy regression in 5.x

### ISSUE THREAD

**Issue by yeahphil** [groue/GRDB.swift#869]
### What did you do?

I've been trying to track down a new crashing bug in an app that recently upgraded from GRDB 4.x -> 5. The culprit is dates stored in this format: `YYYY-MM-DDTHH:MM:SS+0000`

I checked out a copy of master, added to the relevant test case, and see that under GRDB 5, trying to use the `.deferredToDate` strategy to decode dates with these formats calls `fatalConversionError()`:
`YYYY-MM-DDTHH:MM:SS+0000`
`YYYY-MM-DDTHH:MM:SSZ`

Parsing dates in those same formats using GRDB 4.14 succeeds. I know these aren't the framework's preferred date formats, but they are valid ISO8601 dates, and it seems like they should parse when using `.deferredToDate` rather than crashing.

Was this an intentional change in GRDB 5, or is it a regression? If a regression, I'd be happy to take a run at a PR fixing it.

### Environment

**GRDB flavor(s):** Regular GRDB
**GRDB version:** 4.14 -> master
**Installation method:** manual
**Xcode version:** 12.1
**Swift version:** 5.3
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 10.15.7 (Catalina)

**Comment by apptekstudios** [comment]
@yeahphil I believe that `.deferredToDate` uses the fail-able initialisers in [this file](https://github.com/groue/GRDB.swift/blob/master/GRDB/Core/Support/Foundation/Date.swift). Might be worth checking if the Calendar configuration matches your format 👍

**Comment by groue** [comment]
Hello @yeahphil,

This is not a regression, but an intended change. I'm sorry it hit you.

The format used by `DatabaseDateDecodingStrategy.deferredToDate` is the [native SQLite date format](https://sqlite.org/lang_datefunc.html#time_strings). It's the default date format used by GRDB (see [Date and DateComponents](https://github.com/groue/GRDB.swift/blob/master/README.md#date-and-datecomponents)). This format is not to be mismatched with ISO8601, although there is an overlap. Time zones, particularly, are missing.

It happens that GRDB 4.14 *looks like* it would parse time zones. But it would actually ignore them:

```swift
try! DatabaseQueue().read { db in
    let dates = try Date.fetchAll(db, sql: """
              SELECT '2020-11-11T08:00:00Z'
        UNION SELECT '2020-11-11T08:00:00+0000'
        UNION SELECT '2020-11-11T08:00:00+0100'
        """)
    // GRDB 4 prints three times the same Date:
    // 2020-11-11 08:00:00 +0000
    // 2020-11-11 08:00:00 +0000
    // 2020-11-11 08:00:00 +0000
    for date in dates { print(date) }
}
```

GRDB 5 is more strict.

Now, `DatabaseDateDecodingStrategy` allows Codable records to specify another format than the default one. Pick your favorite: https://groue.github.io/GRDB.swift/docs/5.1/Enums/DatabaseDateDecodingStrategy.html

**Comment by yeahphil** [comment]
Thanks for the response and explanation, @groue.

It totally makes sense to go for stricter behavior for correctness. But, I would also note that the sqlite date format docs you linked to do describe an optional trailing Z or timezone info. In my case, a trailing 'Z' in some dates combined with the philosophy of crashing rather than accepting bad data made for a real surprise!

Would you accept a documentation PR? I went looking through the docs before opening this issue and couldn't find a clear answer on what behavior to expect. A couple lines in DatabaseDateDecodingStrategy's docs would have cleared it up quickly.

PS -- thanks for all your work on GRDB, its philosophy and attention to detail have mainly made it a pleasure to use.

**Comment by groue** [comment]
> But, I would also note that the sqlite date format docs you linked to do describe an optional trailing Z or timezone info.

😳 You're right, thank you!

> Formats 2 through 10 may be optionally followed by a timezone indicator of the form "[+-]HH:MM" or just "Z". The date and time functions use UTC or "zulu" time internally, and so the "Z" suffix is a no-op.

OK there's no ambiguity, we need to support that. Precisely, we need to *parse* this format. I'm not sure generating it is needed.

Also note that `[+-]HHMM` and `[+-]HH` are not supported (unlike ISO8601 and your `+0000` example).

> Would you accept a documentation PR?

Doc is updated when code changes, or users ask for clarifications. This time we need a code update in [`SQLiteDateParser`](https://github.com/groue/GRDB.swift/blob/v5.1.0/GRDB/Core/Support/Foundation/SQLiteDateParser.swift) which parses both `Date` and `DateComponent`.

> PS -- thanks for all your work on GRDB, its philosophy and attention to detail have mainly made it a pleasure to use.

Please be assured issues like this one are an essential part of this lib :-) Thanks!

**Comment by groue** [comment]
@yeahphil, I believe you could already solve your issue with the proper `DatabaseDateDecodingStrategy`. I'm closing this in favor of #872.

**Comment by groue** [comment]
Shipped in v5.2.0!
